server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -49,27 +66,85 @@
                 executor, request, listener, shardsIts, timeProvider, clusterStateVersion, task,
                 searchPhaseController.newSearchPhaseResults(task.getProgressListener(), request, shardsIts.size()),
                 request.getMaxConcurrentShardRequests(), clusters);
+        this.topDocsSize = getTopDocsSize(request);
+        this.trackTotalHitsUpTo = request.resolveTrackTotalHitsUpTo();
+        this.hasPrimaryFieldSort = FieldSortBuilder.hasPrimaryFieldSort(request.source());
         this.searchPhaseController = searchPhaseController;
         this.progressListener = task.getProgressListener();
-        final SearchProgressListener progressListener = task.getProgressListener();
         final SearchSourceBuilder sourceBuilder = request.source();
         progressListener.notifyListShards(progressListener.searchShards(this.shardsIts),
             sourceBuilder == null || sourceBuilder.size() != 0);
     }
 
     protected void executePhaseOnShard(final SearchShardIterator shardIt, final ShardRouting shard,
                                        final SearchActionListener<SearchPhaseResult> listener) {
+
+        ShardSearchRequest request = rewriteShardSearchRequest(super.buildShardSearchRequest(shardIt));
         getSearchTransport().sendExecuteQuery(getConnection(shardIt.getClusterAlias(), shard.currentNodeId()),
-            buildShardSearchRequest(shardIt), getTask(), listener);
+            request, getTask(), listener);
     }
 
     @Override
     protected void onShardGroupFailure(int shardIndex, Exception exc) {
         progressListener.notifyQueryFailure(shardIndex, exc);
     }
 
+    @Override
+    protected void onShardResult(SearchPhaseResult result, SearchShardIterator shardIt) {
+        try {
+            QuerySearchResult queryResult = result.queryResult();
+            if (queryResult.isNull() == false
+                    && hasPrimaryFieldSort
+                    && queryResult.topDocs().topDocs instanceof TopFieldDocs) {
+                mergeTopDocs((TopFieldDocs) result.queryResult().topDocs().topDocs, result.getShardIndex());
+            }
+        } catch (Exception exc) {
+            getLogger().error("Failed to merge response", exc);
```

Comment:
I don't think we should ignore errors here, what if this left bottomTopDocs in an inconsistent state?

Reply:
Agreed, this was a leftover from a debug session ;).